### ISSUE TITLE
Add ignore: ["custom-elements"] option to selector-type-no-unknown

### ISSUE THREAD

**Issue by abdonrd** [stylelint/stylelint#2348]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

I want to propose add support to [Custom Elements](https://www.w3.org/TR/custom-elements/).

I add the [specification reference to valid Custom Elements name](https://www.w3.org/TR/custom-elements/#valid-custom-element-name).

> Which rule, if any, is this issue related to?

`selector-type-no-unknown`

> What CSS is needed to reproduce this issue?

```css
my-custom-element {}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
  "rules": {
    "selector-type-no-unknown": true
  }
}
```

> Which version of stylelint are you using?

`7.8.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI and VSCode plugin.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
Unexpected unknown type selector "my-custom-element" (selector-type-no-unknown)
```

**Comment by hudochenkov** [comment]
I don't think this is a good idea. There will be nothing to check by this rule if it allows every custom element. You can just disable rule and receive same result.

**Comment by davidtheclark** [comment]
@abdonrd I think this is a perfect use-case for a [plugin](https://stylelint.io/developer-guide/plugins/). If the spec gains traction, browsers start implementing and people start relying on custom elements — at that time we would reconsider inclusion of the rule in stylelint core.

**Comment by abdonrd** [comment]
@hudochenkov if I disable the rule, we lose the unknown check for non custom elements. Also, maybe we can check the valid Custom Elements name.

**Comment by abdonrd** [comment]
@davidtheclark I think the spec already has traction. It is landed in Chrome, Opera, Safari. And active developing in Firefox.

<img width="1038" alt="screen shot 2017-02-14 at 13 27 24" src="https://cloud.githubusercontent.com/assets/1007051/22928957/5fe68f36-f2b9-11e6-8309-3caa0a9f59f2.png">

**Comment by hudochenkov** [comment]
There is no way to distinguish a custom element from typo in a standard element. E. g. `diva`, it's could be a `div` with typo, or it could be a custom element `diva`. If this rule will support custom elements, it will never catch typos in standard elements.

> I think this is a perfect use-case for a plugin.

I agree with David.

**Comment by jeddy3** [comment]
> Also, maybe we can check the valid Custom Elements name.

@abdonrd Can you expand on this bit please? Do you mean checking for at least one hyphen or is there more to it?

> if I disable the rule, we lose the unknown check for non custom elements.

Can you use the [`ignoreTypes` regex option](https://stylelint.io/user-guide/rules/selector-type-no-unknown/#optional-secondary-options) here?

**Comment by abdonrd** [comment]
@hudochenkov `diva` can't be a Custom Element. A CE needs to:

- Contain a hyphen.
- Starts with lowercase ASCII letter.
- Not contain any uppercase ASCII letters.

**Comment by evilebottnawi** [comment]
@abdonrd best practice for custom element use prefix `x-*` (or you own prefix), use ` ignoreTypes` regex as say @jeddy3 .

**Comment by abdonrd** [comment]
@jeddy3 sure! Check https://github.com/stylelint/stylelint/issues/2348#issuecomment-279697932

And yes! Now I'm using `ignoreTypes`.

**Comment by abdonrd** [comment]
Wow! Thanks *Stylelint Team* for the quick responses! ⚡️ 👏

**Comment by evilebottnawi** [comment]
@abdonrd btw,
1. Some svg tags can contain hyphen https://developer.mozilla.org/en-US/docs/Web/SVG/Element
2. and 3. you can write `<INPUT NAME="TEST">` and it is valid html
Also custom elements can be non `ASCII letters`
```shell
name must match the PotentialCustomElementName production:

PotentialCustomElementName ::=
[a-z] (PCENChar)* '-' (PCENChar)*
PCENChar ::=
"-" | "." | [0-9] | "_" | [a-z] | #xB7 | [#xC0-#xD6] | [#xD8-#xF6] | [#xF8-#x37D] | [#x37F-#x1FFF] | [#x200C-#x200D] | [#x203F-#x2040] | [#x2070-#x218F] | [#x2C00-#x2FEF] | [#x3001-#xD7FF] | [#xF900-#xFDCF] | [#xFDF0-#xFFFD] | [#x10000-#xEFFFF]
```

**Comment by web-padawan** [comment]
After upcoming Safari 10.1 release, there will be two independent CE v1 implementations and I hope you guys will re-consider implementing this. Having Shadow DOM v1 selectors already implemented, this should be enough for linting projects using Web Components.

**Comment by evilebottnawi** [comment]
@web-padawan 
> best practice for custom element use prefix x-* (or you own prefix)

**Comment by web-padawan** [comment]
@evilebottnawi using a prefix is not a best practice, but a requirement. Actually I use the following config:
```
        "selector-type-no-unknown": [ true, {
            "ignoreTypes": ["/^[a-zA-Z]([a-zA-Z0-9]*-[a-zA-Z0-9]+)+/"],
        } ]
```
This should work unless domeone wants to use emoji in CE names 😄

**Comment by evilebottnawi** [comment]
@web-padawan some svg selectors use `-` and i get error on this selectors with you regex

**Comment by evilebottnawi** [comment]
@web-padawan @stylelint/core 
Theoretically we can try this solution (`isCustomElement.js`):
1. Starting with lowercase ASCII letter.
2. Do not contain any uppercase ASCII letters
2. Always contain `-`.
3. It is not `SVG` and `MathML` and `HTML` tags.
Base:
```
NOTE
These requirements ensure a number of goals for valid custom element names:

They start with a lowercase ASCII letter, ensuring that the HTML parser will treat them as tags instead of as text.

They do not contain any uppercase ASCII letters, ensuring that the user agent can always treat HTML elements ASCII-case-insensitively.

They contain a hyphen, used for namespacing and to ensure forward compatibility (since no elements will be added to HTML, SVG, or MathML with hyphen-containing local names in the future).

They can always be created with createElement() and createElementNS(), which have restrictions that go beyond the parser's.

Apart from these restrictions, a large variety of names is allowed, to give maximum flexibility for use cases like <math-α> or <emotion-😍>.
```
I can help with this, if all @stylelint/core accept this issue

**Comment by davidtheclark** [comment]
Thanks @abdonrd and @web-padawan for pointing out that this spec is further along than I had suspected. 

@evilebottnawi I think your plan for the utility sounds good. We can then use it to always ignore custom elements in this rule. Okay?

**Comment by evilebottnawi** [comment]
@davidtheclark yep, ignore custom elements. if someone wants to check custom selectors on unknown (or known) plugin welcome :smile:

**Comment by jeddy3** [comment]
> We can then use it to always ignore custom elements in this rule. Okay?

Yes, let's make this the default behaviour of the rule.

**Comment by evilebottnawi** [comment]
btw, @jeddy3 @davidtheclark maybe refactored labels `syntax: cssnext` (add new), `syntax: scss`, `syntax: less`, syntax: sugarss` and etc.

**Comment by abdonrd** [comment]
Thanks Stylelint Team! Glad to be able to help. :)

BTW, if we use unicode, this regex covers all the [`PotentialCustomElementName`](https://www.w3.org/TR/custom-elements/#prod-potentialcustomelementname):

`^[a-z](-|\.|[0-9]|_|[a-z]|\u{B7}|[\u{C0}-\u{D6}]|[\u{D8}-\u{F6}]|[\u{F8}-\u{37D}]|[\u{37F}-\u{1FFF}]|[\u{200C}-\u{200D}]|[\u{203F}-\u{2040}]|[\u{2070}-\u{218F}]|[\u{2C00}-\u{2FEF}]|[\u{3001}-\u{D7FF}]|[\u{F900}-\u{FDCF}]|[\u{FDF0}-\u{FFFD}]|[\u{10000}-\u{EFFFF}])*-(-|\.|[0-9]|_|[a-z]|\u{B7}|[\u{C0}-\u{D6}]|[\u{D8}-\u{F6}]|[\u{F8}-\u{37D}]|[\u{37F}-\u{1FFF}]|[\u{200C}-\u{200D}]|[\u{203F}-\u{2040}]|[\u{2070}-\u{218F}]|[\u{2C00}-\u{2FEF}]|[\u{3001}-\u{D7FF}]|[\u{F900}-\u{FDCF}]|[\u{FDF0}-\u{FFFD}]|[\u{10000}-\u{EFFFF}])*$`

**Comment by evilebottnawi** [comment]
@abdonrd this regex good for rule `validate-custom-element`, but now in `stylelint` we don't have `validate` rules

**Comment by davidtheclark** [comment]
@evilebottnawi But if we only want to ignore *valid* custom elements — and consider anything else an invalid type selector — this is what we need, right?

**Comment by evilebottnawi** [comment]
@davidtheclark i can do PR, this we can discussion there https://github.com/stylelint/stylelint/pull/2366

**Comment by evilebottnawi** [comment]
Done: https://github.com/stylelint/stylelint/pull/2366

**Comment by FDiskas** [comment]
missing documentation
